tests/robustness/model/replay.go

```diff
@@ -19,82 +19,86 @@ import (
 	"strings"
 )
 
-func NewReplay(eventHistory []PersistedEvent) *EtcdReplay {
-	var lastEventRevision int64 = 1
-	for _, event := range eventHistory {
-		if event.Revision > lastEventRevision && event.Revision != lastEventRevision+1 {
-			panic("Replay requires a complete event history")
+func NewReplay(persistedRequests []EtcdRequest) *EtcdReplay {
+	state := freshEtcdState()
+	revisionToEtcdState := []EtcdState{state, state}
+	var events []PersistedEvent
+	for _, request := range persistedRequests {
+		newState, response := state.Step(request)
+		if state.Revision != newState.Revision {
+			revisionToEtcdState = append(revisionToEtcdState, newState)
+		}
+		for _, e := range toWatchEvents(&state, request, response) {
+			events = append(events, e)
 		}
-		lastEventRevision = event.Revision
+		state = newState
 	}
 	return &EtcdReplay{
-		eventHistory: eventHistory,
+		revisionToEtcdState: revisionToEtcdState,
+		Events:              events,
 	}
 }
 
 type EtcdReplay struct {
-	eventHistory []PersistedEvent
-
-	// Cached state and event index used for it's calculation
-	cachedState       *EtcdState
-	eventHistoryIndex int
+	revisionToEtcdState []EtcdState
+	Events              []PersistedEvent
 }
 
-func (r *EtcdReplay) StateForRevision(revision int64) (EtcdState, error) {
-	if revision < 1 {
-		return EtcdState{}, fmt.Errorf("invalid revision: %d", revision)
-	}
-	if r.cachedState == nil || r.cachedState.Revision > revision {
-		r.reset()
-	}
+func (r *EtcdReplay) StateForRevision(revision int64) EtcdState {
+	return r.revisionToEtcdState[revision]
+}
```

Comment:
Maybe its worth adding a sanity check here for the revision that is passed in. While we append to the `revisionToEtcdState` slice only when the revision changes (and changes means a `++` happens in the robustness test logic somewhere), just in case there is a discrepancy somehow when we pass in the revision via `request.Range.Revision` in `validateSerializableOperation`.

Reply:
Could you propose what kind of validation you have in mind.

Reply:
Done